src/EntityComponentManager_TEST.cc

```diff
@@ -2009,6 +2013,191 @@ TEST_P(EntityComponentManagerFixture, SetChanged)
+TEST_P(EntityComponentManagerFixture, SerializedStateMsgAfterRemoveComponent)
+{
+  // Create entity
+  Entity e1 = manager.CreateEntity();
+  manager.CreateComponent<IntComponent>(e1, IntComponent(123));
+  auto e1c1 =
+    manager.CreateComponent<DoubleComponent>(e1, DoubleComponent(0.0));
+  auto e1c2 =
+    manager.CreateComponent<StringComponent>(e1, StringComponent("int"));
+
+  manager.RemoveComponent(e1, e1c1);
+  manager.RemoveComponent(e1, e1c2);
+
+  // Serialize into a message
+  msgs::SerializedState stateMsg;
+  stateMsg = manager.State();
+
+  // Check message
+  {
+    auto entityMsg = stateMsg.entities(0);
+
+    // First component
+    const auto &e1c0Msg = entityMsg.components(0);
+    EXPECT_FALSE(e1c0Msg.remove());
+
+    // Second component
+    const auto &e1c1Msg = entityMsg.components(1);
+    EXPECT_TRUE(e1c1Msg.remove());
+
+    // Third component
+    const auto &e1c2Msg = entityMsg.components(2);
+    EXPECT_TRUE(e1c2Msg.remove());
+  }
+
+  // Check that removed components don't exist anymore after clearing them
+  manager.RunClearRemovedComponents();
+  msgs::SerializedState newStateMsg;
+  newStateMsg = manager.State();
+
+  // Check message
+  {
+    auto entityMsg = newStateMsg.entities(0);
+    EXPECT_EQ(1, entityMsg.components_size());
+
+    // First component
+    const auto &e1c0Msg = entityMsg.components(0);
+    EXPECT_FALSE(e1c0Msg.remove());
+  }
+}
+
+//////////////////////////////////////////////////
+TEST_P(EntityComponentManagerFixture, RemovedComponentsSyncBetweenServerAndGUI)
+{
+  // Simulate the GUI's ECM
+  EntityCompMgrTest guiManager;
+
+  // Create entity
+  Entity e1 = manager.CreateEntity();
+  manager.CreateComponent<IntComponent>(e1, IntComponent(123));
+  auto e1c1 =
+    manager.CreateComponent<DoubleComponent>(e1, DoubleComponent(0.0));
+  auto e1c2 =
+    manager.CreateComponent<StringComponent>(e1, StringComponent("int"));
+
+  // Serialize server ECM into a message
+  msgs::SerializedStateMap stateMsg;
+  manager.State(stateMsg);
+
+  // Set GUI's ECM and serialize into a message
+  guiManager.SetState(stateMsg);
+  msgs::SerializedStateMap guiStateMsg;
+  guiManager.State(guiStateMsg);
+
+  // Check sync message
+  {
+    auto iter = guiStateMsg.entities().find(e1);
+    const auto &e1Msg = iter->second;
+    auto compIter = e1Msg.components().begin();
+
+    // First component
+    const auto &e1c2Msg = compIter->second;
+    compIter++;
+    EXPECT_FALSE(e1c2Msg.remove());
+
+    // Second component
+    const auto &e1c1Msg = compIter->second;
+    compIter++;
+    EXPECT_FALSE(e1c1Msg.remove());
+
+    // Third component
+    const auto &e1c0Msg = compIter->second;
+    EXPECT_FALSE(e1c0Msg.remove());
+  }
+
+  // Remove components and synchronize again
+  manager.RemoveComponent(e1, e1c1);
+  manager.RemoveComponent(e1, e1c2);
+
+  msgs::SerializedStateMap newStateMsg;
+  manager.State(newStateMsg);
+
+  guiManager.SetState(newStateMsg);
```

Comment:
Before calling `guiManager.State()`,  it would be good to check if the components are not there via the `Component` API

```suggestion
  EXPECT_TRUE(nullptr != guiManager.Component<IntComponent>(e1));
  EXPECT_TRUE(nullptr != guiManager.Component<DoubleComponent>(e1));
  EXPECT_TRUE(nullptr != guiManager.Component<StringComponent>(e1));
  guiManager.SetState(newStateMsg);
  EXPECT_TRUE(nullptr != guiManager.Component<IntComponent>(e1));
  EXPECT_TRUE(nullptr == guiManager.Component<DoubleComponent>(e1));
  EXPECT_TRUE(nullptr == guiManager.Component<StringComponent>(e1));
  
```

Reply:
2f3cd6e